[ARM, AArch64] Fix passing of structures with aligned base classes

hvdijk opened the pull request:
RecordLayout::UnadjustedAlignment was documented as "Maximum of the alignments of the record members in characters", but RecordLayout::getUnadjustedAlignment(), which just returns UnadjustedAlignment, was documented as getting "the record alignment in characters, before alignment adjustement." These are not the same thing: the former excludes alignment of base classes, the latter takes it into account. ItaniumRecordLayoutBuilder::LayoutBase was setting it according to the former, but the AAPCS calling convention handling, currently the only user, relies on it being set according to the latter.

Fixes #135551.

hvdijk commented:
Not entirely sure who to ask for reviews, adding @rjmccall for the Itanium C++ ABI aspect as it interacts with the AAPCS ABIs, and @efriedma-quic who approved https://reviews.llvm.org/D46013 originally.

efriedma-quic commented on clang/lib/AST/RecordLayoutBuilder.cpp:
```diff
@@ -1302,6 +1302,7 @@ ItaniumRecordLayoutBuilder::LayoutBase(const BaseSubobjectInfo *Base) {
     setSize(std::max(getSize(), Offset + Layout.getSize()));
 
   // Remember max struct/class alignment.
+  UnadjustedAlignment = std::max(UnadjustedAlignment, PreferredBaseAlign);
```
Are we supposed to use the alignment of the base class, or the natural alignment of the base class?  I'd like to see testcases for both, and information about how gcc and msvc interpret that question.

Maybe also open a bug report at https://github.com/ARM-software/abi-aa/issues for a spec clarification; Arm is usually pretty responsive there.

------

I would also ask if you want BaseAlign vs. PreferredBaseAlign here, but as far as I can tell, they should be the same on AArch64.

hvdijk replied:
The test case I added has a base class where the alignment is not the natural alignment; GCC uses the alignment here, not the natural alignment, and Compiler Explorer shows that MSVC agrees with GCC: https://godbolt.org/z/4qTY88EW1. The existing `arm-vfp16-arguments2.cpp` test that is updated covers the case where the alignment and natural alignment of the base class are the same.

Thanks for the pointer to get spec clarification. I'll wait for you to confirm whether knowing that GCC and MSVC agree, you'd still like spec clarification; I can ask there tomorrow if you like.

`BaseAlign` vs `PreferredBaseAlign` seems to only be relevant on AIX, unless I am missing something? But thanks, `BaseAlign` is probably better to use, the existing `UnadjustedAlignment = std::max(UnadjustedAlignment, FieldAlign);` also doesn't use `PreferredAlign`. I'll make that change tomorrow when I can confirm that it works as expected.

rjmccall replied:
Does MSVC actually use the ARM ABI? Windows usually just follows its own rules. Ah, well, doesn't really matter.

It seems natural that this would use the alignment of the base class rather than its natural alignment, since that amounts to a consistent rule for all subobjects. I mean, it doesn't *really* make sense to only ignore non-natural alignment at the outermost level, but frankly this is a silly rule to begin with. Every justification for honoring non-natural alignment at the subobject level doubles as an argument for honoring non-natural alignment at the aggregate level, which is the most sensible thing to do, and yet here we are.

Normally, whenever we're talking about base class alignment, we have to be very careful to pick correctly between base-subobject alignment and complete-object alignment. This should probably be base-subobject alignment, but I don't think it's possible to detect a difference because any class with virtual bases is not trivially-copyable and so will always be passed indirectly.

Darwin is just going to have to not follow the ARM spec here. We will not consider an ABI break for something like this.

efriedma-quic replied:
I'm pretty sure Darwin triples opted out of the whole composite alignment thing in 20208cc046c206a1436edb714e94bd52a0bc8572 , so this shouldn't have any effect there.  And... I guess we opted-out Windows too; not sure if that was intentional, since it wasn't discussed at the time, but it seems to match MSVC, so I guess that's fine.

rjmccall replied:
There should at least be a test case for it.

hvdijk replied:
Tested that this doesn't affect Darwin and Windows, but for Windows, that surprises me. Clang matches MSVC for AArch64 in *not* following AAPCS64 on Windows, despite the fact that Microsoft's [Overview of ARM64 ABI conventions > Parameter Passing](https://learn.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=msvc-170#parameter-passing) says "For non-variadic functions, the Windows ABI follows the rules specified by ARM for parameter passing. These rules are excerpted directly from the Procedure Call Standard for the AArch64 Architecture: [...]"

rjmccall replied:
Not too surprising, I hope. I have to confess that Apple used to have a similar statement in our x86_64 ABI documentation until one of our divergences was pointed out (by someone from Microsoft, coincidentally), at which point we rolled up our sleeves and [documented the rest of them](https://developer.apple.com/documentation/xcode/writing-64-bit-intel-code-for-apple-platforms).

hvdijk replied:
From what I can see from the Clang code, for non-HFAs/HVAs, yes, the alignment of the composite type is used rather than the natural alignment, however for HFAs/HVAs, the alignment of the element type is used without regard to the alignment of the composite type. However, this does not seem like it makes any difference because Darwin doesn't have that rule that 16-byte-aligned types are allocated to even-numbered registers anyway, https://godbolt.org/z/qhbYo9c8M

efriedma-quic replied:
If you pass arguments on the stack (use up all the registers), the alignment matters even on Apple... for non-HFA types, at least.  For HFAs, apparently it doesn't matter.  And MSVC has its own funny take on the rules for AArch64: it will pass a misaligned NEON vector on the stack for HVAs containing NEON vectors.

It would be nice to have testcases for all the combinations of Windows/OSX/Linux, register/stack, integer/HFA/HVA, and outer/member alignment... but I guess it doesn't need to be in this patch.  So I'm okay with this going in as-is.

rjmccall replied:
I think it would be better to get that testing in this patch; it really should've been part of the original composite alignment patch, because the fact that some targets explicitly don't use it is a major aspect of the work.

hvdijk replied:
> If you pass arguments on the stack (use up all the registers), the alignment matters even on Apple... for non-HFA types, at least.

Ah, of course, thanks for pointing that out.

> I think it would be better to get that testing in this patch; it really should've been part of the original composite alignment patch, because the fact that some targets explicitly don't use it is a major aspect of the work.

I get where you're coming from but `aapcs64-align.cpp` is specifically for testing AAPCS64 alignment, I don't think it would be right to extend that to checking that non-AAPCS64 targets *don't* use the AAPCS64 ABI. We could possibly add this to the `clang/test/CodeGen/AArch64/args*` tests, I'll have a look.

hvdijk replied:
Done, I've moved the relevant tests from `aapcs64-align.cpp` to `AArch64/args.cpp`. It required small changes because the former tests with `-O2`, the latter without, but nothing that impacts what is being tested.

efriedma-quic reviewed (approved):
LGTM with the PreferredBaseAlign thing fixed.

rjmccall reviewed (changes requested):
Flagging as changes requested to at least verify that this doesn't change ABI for Darwin.

hvdijk committed:
[ARM, AArch64] Fix passing of structures with aligned base classes

RecordLayout::UnadjustedAlignment was documented as "Maximum of the
alignments of the record members in characters", but
RecordLayout::getUnadjustedAlignment(), which just returns
UnadjustedAlignment, was documented as getting "the record alignment in
characters, before alignment adjustement." These are not the same thing:
the former excludes alignment of base classes, the latter takes it into
account. ItaniumRecordLayoutBuilder::LayoutBase was setting it according
to the former, but the AAPCS calling convention handling, currently the
only user, relies on it being set according to the latter.

Alignment tests in aapcs64-align.cpp are moved to AArch64/args.cpp so
that we can test that this change is not applied to Apple.

Fixes #135551.

rjmccall reviewed (approved):
Thanks, that addresses my testing concern.

hvdijk committed:
Merge branch 'main' into fix-135551
